File: seminar_2/task_2/task_2.py
# вторая задачка

class Node(object):
    def __init__(self, data=None):
        self.data = data
        self.next = None
        
class Stack(object):
    def __init__(self):
        self.top = None
    
    def push(self, data):
        new_Node = Node(data)
        
        if self.top is None:
            self.top = new_Node
            return
        
        new_Node.next = self.top
        self.top = new_Node

        
    def pop(self):
        if not self.top:
            return None
        
        top = self.top
        if self.top.next is not None:
            self.top = self.top.next
        else:
            self.top = None
            
        return top.data    
    #def show_stack(self):
        

def isPalindrome(s):
    st = Stack()
    
    for char in s:
        st.push(char)
    
    for char in s:
        
        if char != st.pop():
            return False
    
    return True

a = 'SHALAHS'

print(isPalindrome(a))

File: seminar_2/task_2/task2.cpp
#include <iostream>
#include <stack>
using namespace std;

bool is_palindrome(string s){
    stack<char> st;
    for(size_t i=0; i<s.size(); i++)
        st.push(s[i]);
    for(size_t i=0; i<s.size(); i++){
        char top = st.top();
        if(top != s[i])
            return false;
        st.pop();}
    return true;
}

int main(void){
    cout<<is_palindrome("")<<endl;
    cout<<is_palindrome("A")<<endl;
    cout<<is_palindrome("AA")<<endl;
    cout<<is_palindrome("AB")<<endl;
    cout<<is_palindrome("ABA")<<endl;
    cout<<is_palindrome("SABAS")<<endl;
    cout<<is_palindrome("SABAJ")<<endl;
    cout<<is_palindrome("ALOOLA")<<endl;
    return 0;
}

